### PR TITLE
[C-API] Define new API for tensor_if

### PR DESCRIPTION
**Pull request by gichan-jang**
Define new API for tensor_if custom register/unregister

Signed-off-by: gichan-jang <gichan2.jang@samsung.com>

Self evaluation:
1. Build test: [ *]Passed [ ]Failed [ ]Skipped
2. Run test: [ *]Passed [ ]Failed [ ]Skipped

### CODE REVIEW

**Review by dongju-chae — APPROVED** [review]
LGTM

File: api/capi/include/nnstreamer.h [hunk 1]
```diff
@@ -238,6 +244,16 @@ typedef void (*ml_pipeline_sink_cb) (const ml_tensors_data_h data, const ml_tens
  */
 typedef void (*ml_pipeline_state_cb) (ml_pipeline_state_e state, void *user_data);
 
+/**
+ * @brief Callback for custom condition of tensor_if
+ * @since_tizen 6.5
+ * @param[out] info The handle of tensors information (cardinality, dimension, and type of given tensor/tensors).
+ * @param[out] data The handle of the tensor output (a single frame. tensor/tensors). Number of tensors is determined by ml_tensors_info_get_count() with the handle 'info'. Note that the maximum number of tensors is 16 (#ML_TENSOR_SIZE_LIMIT).
+ * @param[out] user_data User application's private data.
+ * @param[out] result Result of the user defined condition
```
Comment (jaeyun-jung) [inline api/capi/include/nnstreamer.h:253]: how about returning int value as result? not result ptr.
int if_custom_cb (info, data, user_data)

Comment (gichan-jang) [reply api/capi/include/nnstreamer.h:253]: How about returning the error code?

**Review by jaeyun-jung — APPROVED** [review]

File: api/capi/include/nnstreamer.h (hunk 1 above)
Comment (myungjoo) [reply api/capi/include/nnstreamer.h:253]: Are you able to handle error returns from this callback?

Comment (myungjoo) [reply api/capi/include/nnstreamer.h:253]: The callback function defined by tensor-if returns TRUE if no error. So, I agree with @gichan-jang 's intention.

Comment (myungjoo) [reply api/capi/include/nnstreamer.h:253]: ```Result of the user defined condition``` 
==>
```Result of the user-defined condition. 0 refers to FALSE and a non-zero value refers to TRUE.```

File: api/capi/include/nnstreamer.h [hunk 2]
```diff
@@ -885,6 +904,43 @@ int ml_pipeline_element_get_property_double (ml_pipeline_element_h elem_h, const
...
 
+/****************************************************
+ ** NNStreamer Pipeline tensor_if Control          **
+ ****************************************************/
+/**
+ * @brief Registers a tensor_if custom callback.
+ * @details If the if-condition is complex and cannot be expressed with tensor_if expressions, you can define custom condition.
+ * @since_tizen 6.5
+ * @remarks If the function succeeds, @a tensor_if handle must be released using ml_pipeline_if_custom_unregister().
+ * @param[in] name The name of tensor_if
```
Comment (myungjoo) [inline api/capi/include/nnstreamer.h:915]: Is this ```name```, the name of tensor-if element? or the name of custom condition?

File: api/capi/include/nnstreamer.h [hunk 3]
```diff
@@ -885,6 +904,43 @@ int ml_pipeline_element_get_property_double (ml_pipeline_element_h elem_h, const
...
+ * @retval #ML_ERROR_NOT_SUPPORTED Not supported.
+ * @retval #ML_ERROR_INVALID_PARAMETER The parameter is invalid, or duplicated name exists.
+ * @retval #ML_ERROR_OUT_OF_MEMORY Failed to allocate required memory to register the custom filter.
+ *
+ * Here is an example of the usage:
+ * @code
+ * TBU
+ * @endcode
+ */
+int ml_pipeline_tensor_if_custom_register (const char *name, ml_pipeline_if_custom_cb cb, void *user_data, ml_pipeline_if_h *if_custom);
```
Comment (myungjoo) [inline api/capi/include/nnstreamer.h:930]: We are registering a custom condition to the process globally. We need to write some notes or warning on this.
(If we register a custom condition "X", this "X" may be referred in any pipelines of the current process)

**Review by myungjoo — CHANGES_REQUESTED** [review]
Need further discussion

File: api/capi/include/nnstreamer.h (hunk 1 above)
Comment (gichan-jang) [reply api/capi/include/nnstreamer.h:253]: Thanks. I will update.

File: api/capi/include/nnstreamer.h (hunk 2 above)
Comment (gichan-jang) [reply api/capi/include/nnstreamer.h:915]: Name of the tensor_if element.

File: api/capi/include/nnstreamer.h (hunk 3 above)
Comment (gichan-jang) [reply api/capi/include/nnstreamer.h:930]: Is it enough to add a comments? or Post debug or warning messages?

File: api/capi/include/nnstreamer.h (hunk 2 above)
Comment (myungjoo) [reply api/capi/include/nnstreamer.h]: Why do  you need the name of tensor_if element?

From tensor_if.h:
```
 *
 * ...
 * // Register custom callback function describing the condition
 * nnstreamer_if_custom_register ("tifx", tensor_if_custom_cb, NULL);
 * ...
 * // Use the condition in a pipeline.
 * // E.g., Pipeline of " ... ! tensor_if compared-value=CUSTOM compared-value-option=tifx then=TENSORPICK then-option=0 else=TENSORPICK else-option=1 ! ... "
 * ...
 * // After everything is done.
 * nnstreamer_if_custom_unregister ("tifx");
 * @endcode
```

Comment (gichan-jang) [reply api/capi/include/nnstreamer.h]: Oh my.. You're right. Updated.

File: api/capi/include/nnstreamer.h [hunk 4]
```diff
@@ -238,6 +244,19 @@ typedef void (*ml_pipeline_sink_cb) (const ml_tensors_data_h data, const ml_tens
  */
 typedef void (*ml_pipeline_state_cb) (ml_pipeline_state_e state, void *user_data);
 
+/**
+ * @brief Callback for custom condition of tensor_if
+ * @since_tizen 6.5
+ * @param[out] info The handle of tensors information (cardinality, dimension, and type of given tensor/tensors).
+ * @param[out] data The handle of the tensor output (a single frame. tensor/tensors). Number of tensors is determined by ml_tensors_info_get_count() with the handle 'info'. Note that the maximum number of tensors is 16 (#ML_TENSOR_SIZE_LIMIT).
```
Comment (myungjoo) [inline api/capi/include/nnstreamer.h:251]: output? not an input?
out --> in ?

File: api/capi/include/nnstreamer.h [hunk 5]
```diff
@@ -238,6 +244,19 @@ typedef void (*ml_pipeline_sink_cb) (const ml_tensors_data_h data, const ml_tens
  */
 typedef void (*ml_pipeline_state_cb) (ml_pipeline_state_e state, void *user_data);
 
+/**
+ * @brief Callback for custom condition of tensor_if
+ * @since_tizen 6.5
+ * @param[out] info The handle of tensors information (cardinality, dimension, and type of given tensor/tensors).
```
Comment (myungjoo) [inline api/capi/include/nnstreamer.h:250]: out --> in ?

File: api/capi/include/nnstreamer.h [hunk 6]
```diff
@@ -238,6 +244,19 @@ typedef void (*ml_pipeline_sink_cb) (const ml_tensors_data_h data, const ml_tens
...
+ * @param[out] data The handle of the tensor output (a single frame. tensor/tensors). Number of tensors is determined by ml_tensors_info_get_count() with the handle 'info'. Note that the maximum number of tensors is 16 (#ML_TENSOR_SIZE_LIMIT).
+ * @param[out] user_data User application's private data.
```
Comment (myungjoo) [inline api/capi/include/nnstreamer.h:252]: out --> in/out

File: api/capi/include/nnstreamer.h (hunk 3 above)
Comment (myungjoo) [reply api/capi/include/nnstreamer.h:933]: It is to be commented: let users know that a registered if-custom can be referred anywhere from the user process.

File: api/capi/include/nnstreamer.h (hunk 4 above)
Comment (again4you) [reply api/capi/include/nnstreamer.h:251]: @myungjoo  Nice Catch! We need to update it.

Comment (gichan-jang) [reply api/capi/include/nnstreamer.h]: Thanks! Updated.
I will upload other related revisions.

File: api/capi/include/nnstreamer.h (hunk 5 above)
Comment (gichan-jang) [reply api/capi/include/nnstreamer.h]: Updated.

File: api/capi/include/nnstreamer.h (hunk 6 above)
Comment (gichan-jang) [reply api/capi/include/nnstreamer.h]: Updated.

**Review by myungjoo — APPROVED** [review]
LGTM

**Comment by myungjoo** [comment]
PTAL: @again4you , @leemgs

File: api/capi/include/nnstreamer.h [hunk 7]
```diff
@@ -238,6 +244,19 @@ typedef void (*ml_pipeline_sink_cb) (const ml_tensors_data_h data, const ml_tens
  */
 typedef void (*ml_pipeline_state_cb) (ml_pipeline_state_e state, void *user_data);
 
+/**
+ * @brief Callback for custom condition of tensor_if
+ * @since_tizen 6.5
+ * @param[in] info The handle of tensors information (cardinality, dimension, and type of given tensor/tensors).
+ * @param[in] data The handle of the tensor output of the pipeline (a single frame. tensor/tensors). Number of tensors is determined by ml_tensors_info_get_count() with the handle 'info'. Note that the maximum number of tensors is 16 (#ML_TENSOR_SIZE_LIMIT).
+ * @param[in/out] user_data User application's private data.
+ * @param[out] Result of the user-defined condition. 0 refers to FALSE and a non-zero value refers to TRUE.
```
Comment (jaeyun-jung) [inline api/capi/include/nnstreamer.h:253]: add missed arg `@param[out] result Result`

File: api/capi/include/nnstreamer.h [hunk 8]
```diff
@@ -238,6 +244,19 @@ typedef void (*ml_pipeline_sink_cb) (const ml_tensors_data_h data, const ml_tens
...
+ * @return @c 0 on success. Otherwise a negative error value.
+ * @retval #ML_ERROR_NONE Successful.
```
Comment (jaeyun-jung) [inline api/capi/include/nnstreamer.h:255]: IMO, this is user callback, so ml error code is useless.

File: api/capi/include/nnstreamer.h [hunk 9]
```diff
@@ -885,6 +904,46 @@ int ml_pipeline_element_get_property_double (ml_pipeline_element_h elem_h, const
...
+ * @param[in] cb The function to be called when the pipeline runs.
+ * @param[in] user_data Private data for the callback. This value is passed to the callback when it's invoked.
+ * @param[out] if_custom The tensor_if handler.
+ * @return @c 0 on success. Otherwise a negative error value.
+ * @retval #ML_ERROR_NONE Successful.
+ * @retval #ML_ERROR_NOT_SUPPORTED Not supported.
+ * @retval #ML_ERROR_INVALID_PARAMETER The parameter is invalid, or duplicated name exists.
+ * @retval #ML_ERROR_OUT_OF_MEMORY Failed to allocate required memory to register the custom filter.
+ * @warning A custom condition of the tensor_if is registered to the process globally.
+ *          If the custom condition "X" is registerd , this "X" may be referred in any pipelines of the current proces.
```
Comment (jaeyun-jung) [inline api/capi/include/nnstreamer.h:925]: typo proces

File: api/capi/include/nnstreamer.h (hunk 7 above)
Comment (gichan-jang) [reply api/capi/include/nnstreamer.h]: Updated.

File: api/capi/include/nnstreamer.h (hunk 8 above)
Comment (gichan-jang) [reply api/capi/include/nnstreamer.h]: Right. ml error code is useless. updated.

File: api/capi/include/nnstreamer.h (hunk 9 above)
Comment (gichan-jang) [reply api/capi/include/nnstreamer.h]: Updated.

File: api/capi/include/nnstreamer.h [hunk 10]
```diff
@@ -885,6 +902,46 @@ int ml_pipeline_element_get_property_double (ml_pipeline_element_h elem_h, const
...
+ * @param[in] cb The function to be called when the pipeline runs.
+ * @param[in] user_data Private data for the callback. This value is passed to the callback when it's invoked.
+ * @param[out] if_custom The tensor_if handler.
+ * @return @c 0 on success. Otherwise a negative error value.
+ * @retval #ML_ERROR_NONE Successful.
+ * @retval #ML_ERROR_NOT_SUPPORTED Not supported.
+ * @retval #ML_ERROR_INVALID_PARAMETER The parameter is invalid, or duplicated name exists.
+ * @retval #ML_ERROR_OUT_OF_MEMORY Failed to allocate required memory to register the custom filter.
+ * @warning A custom condition of the tensor_if is registered to the process globally.
+ *          If the custom condition "X" is registerd , this "X" may be referred in any pipelines of the current process.
```
Comment (again4you) [inline api/capi/include/nnstreamer.h:923]: * registerd -> register**e**d

**Review by again4you — COMMENTED** [review]
Except for simple typos, LGTM!

* `Defube new` -> `Define new` on commit message.

File: api/capi/include/nnstreamer.h (hunk 10 above)
Comment (gichan-jang) [reply api/capi/include/nnstreamer.h]: Thanks. Updated..!

**Review by again4you — APPROVED** [review]
LGTM!

**Pull request merged**